runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/job/ReferenceRunnerJobService.java

```diff
@@ -201,6 +204,36 @@ public void getState(
     responseObserver.onCompleted();
   }
 
+  @Override
+  public void getStateStream(
+      GetJobStateRequest request, StreamObserver<GetJobStateResponse> responseObserver) {
+    LOG.trace("{} {}", GetJobStateRequest.class.getSimpleName(), request);
+    String invocationId = request.getJobId();
+    try {
+      Thread.sleep(WAIT_MS);
+      Enum state = jobStates.getOrDefault(request.getJobId(), Enum.UNRECOGNIZED);
+      responseObserver.onNext(GetJobStateResponse.newBuilder().setState(state).build());
+      while (Enum.RUNNING.equals(state)) {
+        Thread.sleep(WAIT_MS);
+        state = jobStates.getOrDefault(request.getJobId(), Enum.UNRECOGNIZED);
+      }
+      responseObserver.onNext(GetJobStateResponse.newBuilder().setState(state).build());
```

Comment:
This seems wrong based on my understanding of onNext. I assumed (based on a quick reading of the documentation for StreamObservers) that onNext would have to be called every time the state enum changed. With the call before and after the while loop, it seems like it would be called twice in succession if the state is initially something other than Enum.RUNNING.

Reply:
Makes sense. And it seems that in the case of it not being in the running state, it's not an issue to republish the state, so there's no performance/correctness issues here.